packages/webr/src/outputconnection.c

```diff
@@ -0,0 +1,101 @@
+#include <R.h>
+#include <Rinternals.h>
+#include <R_ext/Connections.h>
+
+# define BUFSIZE 10000
+struct output_con_data {
+  SEXP output;
+  char buf[BUFSIZE];
+  char *line, *cur;
+};
+
+static Rboolean output_open(Rconnection con) {
+  con->isopen = TRUE;
+  return TRUE;
+}
+
+static void output_close(Rconnection con) {
+  con->isopen = FALSE;
+}
+
+static void output_destroy(Rconnection con) {
+  if(con->private) free(con->private);
+}
+
+static int output_vfprintf(Rconnection con, const char *format, va_list ap) {
+  struct output_con_data* data = con->private;
+  int res;
+  char *p, *q;
+  size_t curlen = data->cur - data->buf;
+
+  va_list aq;
+  va_copy(aq, ap);
+  res = vsnprintf(data->cur, BUFSIZE - curlen, format, aq);
+  va_end(aq);
+
+  if (res < 0) {
+    data->cur[0] = '\0';
+  } else if(res >= BUFSIZE - curlen) {
+	  warning("printing of extremely long output is truncated");
+    data->buf[BUFSIZE-1] = '\0';
+	  res = BUFSIZE - curlen - 1;
+  }
+  data->cur += res < 0 ? 0 : res;
+
+  // Check for newlines in buffer and copy completed lines to output
+  for (p = data->line; p < data->cur; p++) {
+    if (*p == '\n') {
+      *p = '\0';
+      const char *names[] = {"type", "data"};
+      SEXP elt = PROTECT(Rf_mkNamed(VECSXP, names));
+      SET_VECTOR_ELT(elt, 0, Rf_mkString(con->description));
+      SET_VECTOR_ELT(elt, 1, Rf_mkChar(data->line));
+
+      SEXP vec = VECTOR_ELT(data->output, 2);
+      int len = Rf_length(vec);
+      vec = PROTECT(Rf_lengthgets(vec, len + 1));
+      SET_VECTOR_ELT(vec, len, elt);
+      SET_VECTOR_ELT(data->output, 2, vec);
+
+      UNPROTECT(2);
+      data->line = p + 1;
+    }
+  }
+
+  // If we're not in the middle of a line, reset to start of buffer
+  if (data->line == data->cur) {
+    data->cur = data->line = data->buf;
+  }
+  return res;
+}
+
+void init_output_connection(Rconnection con, SEXP out) {
+  con->open = &output_open;
+  con->close = &output_close;
+  con->vfprintf = &output_vfprintf;
+  con->destroy = &output_destroy;
+
+  struct output_con_data* data = malloc(sizeof (struct output_con_data));
+  data->output = out;
+  data->cur = data->line = data->buf;
+  con->private = data;
+}
+
+SEXP new_output_connections() {
+  Rconnection out_con_stdout;
+  Rconnection out_con_stderr;
+
+  const char *names[] = {"stdout", "stderr", "vec"};
```

Comment:
Needs a `""` sentinel.

```suggestion
  const char *names[] = { "stdout", "stderr", "vec", "" };
```